### DISCUSSION TITLE
Noodle: GuitabSDK vs. MailSender

### DISCUSSION THREAD

**Discussion by mfmarano** [dds-jv/foro#403] [Consultas de objetos] [answered]
Buenas, quería consultar algo del [enunciado de Noodle](https://docs.google.com/document/d/1-Qpv38kB29lNuzIi88TkTg4LGbO_iNijoVesZyvG5wM/edit).

![image](https://github.com/dds-jv/foro/assets/82288948/d1b04075-978b-439b-a177-d837b5086ef2)

## Qué son
¿Es a propósito que se modele a `GuitabSDK` como una interfaz y a `MailSender` como una clase? ¿Es esto porque `GuitabSDK` es una interfaz externa y `MailSender` una biblioteca?
- yo interpretaba que ambas son interfaces externas

## Cómo se usan
si bien entiendo que para interactuar con sistemas externos debemos adaptarlos, en la solución de Noodle veo que se instancian directamente en los interesados de un Grupo (o sea, las clases que implementan `Tarea`).
De esta manera,
1. ¿los interesados de un `Grupo` tendrían también el rol de adaptar estas interfaces externas?
2. ¿podría haber sido una alternativa la siguiente?:
a. crear una interfaz `RepoGit`, una clase `Guitab` que implemente `RepoGit` y se inyecte como dependencia `GuitabSDK`
b. crear una interfaz `Notificador`, una clase `Mailer` que implemente `Notificador` y se inyecte como dependencia `MailSender`

**Comment by flbulgarelli** [comment]
¡Hola!

> ¿Es a propósito que se modele a GuitabSDK como una interfaz y a MailSender como una clase? ¿Es esto porque GuitabSDK es una interfaz externa y MailSender una biblioteca?
>
>    yo interpretaba que ambas son interfaces externas

Correcto. Ambas son interfaces externas, independientemente de si a nivel modelado en objetos son clases o interfaces. 

>  en la solución de Noodle veo que se instancian directamente en los interesados de un Grupo (o sea, las clases que implementan Tarea).

Es cierto, para no volver al texto más extenso no se mencionó la adaptación de interfaces, pero el objetivo es que dichas interfaces sí sean adaptadas (lo acabo de agregar como un comentario en la solución). Por lo tanto tu propuesta de introducir dos _adapters_ para cada uno de estos servicios es correcta. 

Más allá de eso, como bien vos señalás...

> ¿los interesados de un Grupo tendrían también el rol de adaptar estas interfaces externas?

... si no se usaran adapters de todas formas los interesados en el grupo estarían de alguna forma oficiando de adaptación, lo cual sería un paliativo a los  problemas de acoplamiento y testeabilidad. Aún así, sugiero que en situaciones como estas uses adapters para mayor seguridad en el parcial.

**Reply by mfmarano** [reply]
Hola, ¡gracias por la respuesta!

Entiendo lo de usar adapters para cada servicio y que cada servicio es una interfaz externa.

Lo que aún no me queda claro es por qué `GuitabSDK` es una interfaz y `MailSender` una clase a nivel modelado en objetos. Y si esto impacta de alguna forma en cómo implemente a estos servicios en el parcial, más allá de que la interacción con estos sea a través de adapters.

**Reply by flbulgarelli** [reply, marked as answer]
En líneas generales no existe **ninguna diferencia** entre que `GuitabSDK` y `MailSender` sean interfaces y clases, respectivamente. De hecho, todo lo contrario: está allá justamente para recordarte que poco importa cómo estén implementados los componentes externos y que a fin de cuentas, son sólo eso: piezas de software de las que no deberías asumir su estabilidad y contra las que querés minimizar el acoplamiento.

Dicho eso, en el mundo real, sí podríamos encontrar diferencias, tanto desde el punto de vista creacional como desde algún pequeño grado de libertad adicional en el diseño que te ofrecería que `GuitabSDK` sea una interfaz a la hora del testing. Pero todo eso se diluye tanto en un parcial como al usar tecnologías mas o menos modernas (por ejemplo, con [PowerMock](https://github.com/powermock/powermock) casi cualquier cosa es _mockeable_) que se vuelve irrelevante para el problema de Noodle. 

En resumen ambos sos componentes externos, sin más.

**Comment by mfmarano** [comment]
Gracias!
Quedó todo super claro 😄
